### PR TITLE
Creating dto, mapper, repository and service for category.

### PR DESCRIPTION
**Pull request by FortuneAwaker**
Category stuff.

### CODE REVIEW

File: src/main/java/by/itech/lab/supplier/service/impl/CategoryService.java [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+    public CategoryService(final CategoryRepository categoryRepository,
+                           final CategoryMapper mapper) {
+        this.categoryRepository = categoryRepository;
+        this.mapper = mapper;
+    }
+
+    public CategoryDto findByCategory(final String categoryName) {
+        Optional<Category> optionalCategory
+          = categoryRepository.findByCategory(categoryName);
```
Comment (svareyko) [inline src/main/java/by/itech/lab/supplier/service/impl/CategoryService.java:30]: wtf?

File: src/main/java/by/itech/lab/supplier/service/impl/CategoryService.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+            category = optional.get();
+        } else {
+            return null;
+        }
+        return mapper.map(category);
+    }
+}
+
+
+
```
Comment (svareyko) [inline src/main/java/by/itech/lab/supplier/service/impl/CategoryService.java:81]: spaces?

File: src/main/java/by/itech/lab/supplier/service/impl/CategoryService.java [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+        return checkOptional(optional);
+    }
+
+    public void delete(final CategoryDto dto) {
+        categoryRepository.delete(mapper.mapCategoryWithId(dto));
+    }
+
+    private CategoryDto checkOptional(final Optional<Category> optional) {
+        Category category;
+        if (optional.isPresent()) {
```
Comment (svareyko) [inline src/main/java/by/itech/lab/supplier/service/impl/CategoryService.java:71]: why just not use something like:
repository.findById(..).orElse(null)
or even simpler add method findById without optional in output?

File: src/main/java/by/itech/lab/supplier/service/impl/CategoryService.java [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+
+    public CategoryDto findByCategory(final String categoryName) {
+        Optional<Category> optionalCategory
+          = categoryRepository.findByCategory(categoryName);
+        return checkOptional(optionalCategory);
+    }
+
+    public CategoryDto create(final CategoryDto dto) throws NotUniqueException {
+        Category category;
+        if (dto.getId() == null) {
```
Comment (svareyko) [inline src/main/java/by/itech/lab/supplier/service/impl/CategoryService.java:36]: Objects.isNull

File: src/main/java/by/itech/lab/supplier/service/impl/CategoryService.java [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+            category = mapper.map(dto);
+        } else {
+            category = mapper.mapCategoryWithId(dto);
+        }
+        CategoryDto created;
+        try {
+            created = mapper.map(categoryRepository.save(category));
+        } catch (PropertyValueException notUnique) {
+            throw new NotUniqueException(notUnique);
```
Comment (svareyko) [inline src/main/java/by/itech/lab/supplier/service/impl/CategoryService.java:45]: not sure that understand why you re-throw that exception

File: src/main/java/by/itech/lab/supplier/service/impl/CategoryService.java [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+        }
+        return created;
+    }
+
+    public List<CategoryDto> readAll() {
+        List<Category> categories = categoryRepository.findAll();
+        List<CategoryDto> dtos = new ArrayList<>();
+        for (Category item: categories
```
Comment (svareyko) [inline src/main/java/by/itech/lab/supplier/service/impl/CategoryService.java:53]: what about streams?

File: src/main/java/by/itech/lab/supplier/service/impl/CategoryService.java [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+             ) {
+            dtos.add(mapper.map(item));
+        }
+        return dtos;
+    }
+
+    public CategoryDto readById(final Long id) {
+        Optional<Category> optional = categoryRepository.findById(id);
+        return checkOptional(optional);
```
Comment (svareyko) [inline src/main/java/by/itech/lab/supplier/service/impl/CategoryService.java:62]: see line 71

File: src/main/java/by/itech/lab/supplier/service/impl/CategoryService.java [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+    public CategoryService(final CategoryRepository categoryRepository,
+                           final CategoryMapper mapper) {
+        this.categoryRepository = categoryRepository;
+        this.mapper = mapper;
+    }
+
+    public CategoryDto findByCategory(final String categoryName) {
+        Optional<Category> optionalCategory
+          = categoryRepository.findByCategory(categoryName);
+        return checkOptional(optionalCategory);
```
Comment (svareyko) [inline src/main/java/by/itech/lab/supplier/service/impl/CategoryService.java:31]: see line 71

File: src/test/java/by/itech/lab/supplier/CategoryTest.java [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+    void checkConnection() {
+        CategoryDto category = new CategoryDto();
+        category.setCategory("Test");
+        category.setTaxRate(20.0);
+        CategoryDto category1 = new CategoryDto();
+        category.setCategory("Test");
+        category.setTaxRate(20.0);
+        service.create(category);
+        service.create(category1);
+        System.out.println(service.readAll());
```
Comment (svareyko) [inline src/test/java/by/itech/lab/supplier/CategoryTest.java:26]: where is assertions? who will read application output?

**Review by svareyko — CHANGES_REQUESTED** [review]

File: src/main/java/by/itech/lab/supplier/service/impl/CategoryService.java (hunk 2 above)
Comment (FortuneAwaker) [reply src/main/java/by/itech/lab/supplier/service/impl/CategoryService.java:81]: done

File: src/main/java/by/itech/lab/supplier/service/impl/CategoryService.java (hunk 4 above)
Comment (FortuneAwaker) [reply src/main/java/by/itech/lab/supplier/service/impl/CategoryService.java:36]: done

File: src/main/java/by/itech/lab/supplier/service/impl/CategoryService.java (hunk 5 above)
Comment (FortuneAwaker) [reply src/main/java/by/itech/lab/supplier/service/impl/CategoryService.java:45]: done

File: src/main/java/by/itech/lab/supplier/service/impl/CategoryService.java (hunk 1 above)
Comment (FortuneAwaker) [reply src/main/java/by/itech/lab/supplier/service/impl/CategoryService.java:30]: done

File: src/main/java/by/itech/lab/supplier/service/impl/CategoryService.java (hunk 3 above)
Comment (FortuneAwaker) [reply src/main/java/by/itech/lab/supplier/service/impl/CategoryService.java:71]: done

File: src/main/java/by/itech/lab/supplier/service/impl/CategoryService.java (hunk 7 above)
Comment (FortuneAwaker) [reply src/main/java/by/itech/lab/supplier/service/impl/CategoryService.java:62]: done

File: src/main/java/by/itech/lab/supplier/service/impl/CategoryService.java (hunk 8 above)
Comment (FortuneAwaker) [reply src/main/java/by/itech/lab/supplier/service/impl/CategoryService.java:31]: done

**Pull request merged**